### ISSUE TITLE
Temporary folder names conflict when running same sorter multiple times in a script.

### ISSUE THREAD

**Issue by Shawn-Guo-CN** [SpikeInterface/spikesorters#59]
When I try to run a same sorter, e.g. tridesclous, in a python script, the programme would crash because that 2 different instances of a same sorter would use the same **Temporary Folder** and the lock of files wouldn't be released until the end of main process. Thus, it becomes impossible to run a sorter with several different hyperparameter settings.

**Comment by samuelgarcia** [comment]
If you run several instance in parralel you have to force **output_folder** by hand because it is None by deafult and None is translated to ```output_folder = 'tmp_' + self.sorter_name``` in basesorter.

**Comment by Shawn-Guo-CN** [comment]
> If you run several instance in parralel you have to force **output_folder** by hand because it is None by deafult and None is translated to `output_folder = 'tmp_' + self.sorter_name` in basesorter.

I was wondering, maybe we could add a timestamp as a postifx of tmp folder names?

**Comment by samuelgarcia** [comment]
For the normal case, I don't like it so much.
It is cool a a naive user to see that a folder with the name of the sorter have been create.
I think for more expert user playing with the argument output_folder is quite easy.
In your case for parameters search I would modify this folder according to the paramters values for instance.

**Comment by Shawn-Guo-CN** [comment]
Yep, what you state definitely holds. Then, we need to change the codes in [Getting Started tutorial](https://spikeinterface.readthedocs.io/en/latest/getting_started/plot_getting_started.html), as the mountainsort4 runs 2 times in it.

**Comment by samuelgarcia** [comment]
In the context of this getting start, I think it is OK to simplify and not give too much details (about ouput_folder for intance). The 2 times are to illustrate that we can change parameters.
To avoid confucion in this example we should sorting_MS4 2 times as output.

@alejoe91 : what do you think ?

**Comment by Shawn-Guo-CN** [comment]
> In the context of this getting start, I think it is OK to simplify and not give too much details (about ouput_folder for intance). The 2 times are to illustrate that we can change parameters.
> To avoid confucion in this example we should sorting_MS4 2 times as output.
> 
> @alejoe91 : what do you think ?

I just worry that it may confuse users if the example script crashes.

**Comment by Shawn-Guo-CN** [comment]
Is the [Getting Started tutorial](https://spikeinterface.readthedocs.io/en/latest/getting_started/plot_getting_started.html) based on out-of-date version? The output of second `run_mountainsort4` is `/tmp/tmpqdf1b_0j/timeseries.hdf5`.

**Comment by Shawn-Guo-CN** [comment]
> ??? Maybe there is a hack on the the folder name in MS itself.
> I will have a look.

After checking file operations in our project, I believe that it is more possible that this problem is caused by sorters themselves. I'm checking the subprocess operations, to see how/where to kill the subprocesses in our programmes.

**Comment by Shawn-Guo-CN** [comment]
> > ??? Maybe there is a hack on the the folder name in MS itself.
> > I will have a look.
> 
> After checking file operations in our project, I believe that it is more probable that this problem is caused by sorters themselves. I'm checking the subprocess operations, to see how/where to kill the subprocesses in our programmes.

Now, I got it. It's the permission problem on Win... We just don't have full I/O priviledge on files created by ourselves...

**Comment by Shawn-Guo-CN** [comment]
I just realised that this issue is bigger and more serious than I thought before. On Windows, when we run some sorters, the temporary folders won't be removed even we set the "delete_output_folder" flag to be `True`. 

The cause of this issue is that the temporary folder keeps being accessed by some sub-processes (which I don't know who they are) as well as they are read-only for user account.

It seems that we find a better way to manage the temporary folders on Windows.

**Comment by alejoe91** [comment]
Hi @Shawn-Guo-CN I don't really know what we can do about this, rather then adding try except everytime we try to delete a folder. 

Which folder are you using? Are you trying to spike sort in parallel?

**Comment by Shawn-Guo-CN** [comment]
Hi @alejoe91 , I'll try to find a solution for this issue. I just want to leave it here to remind myself. 

I use the standard tmp folder, e.g. 'tmp_klusta', and I run just one sorter. But the BaseSorter cannot remove the tmporary folder. It's caused by Windows file privilege management strategy, I'm still looking for a feasible solution.
